### PR TITLE
SOLR-15762: Detect invalid filter cache configuration during join

### PR DESCRIPTION
**Pull request by thomaswoeckinger**

### CODE REVIEW

**Comment by thomaswoeckinger** [comment]
@madrob Plz review

**Review by madrob — APPROVED** [review]
LGTM will wait for precommit to run

**Comment by dsmiley** [comment]
I'm running this test and it failed the first time (meaning the expected exception wasn't thrown); the second is hanging on me still.

File: solr/core/src/java/org/apache/solr/search/JoinQParserPlugin.java [hunk 1]
```diff
@@ -587,17 +587,27 @@ public DocSet getDocSetEnumerate() throws IOException {
...
-              DocSet toTermSet = toSearcher.getDocSet(toDeState);
-              resultListDocs += toTermSet.size();
-              if (resultBits != null) {
-                toTermSet.addAllTo(new BitDocSet(resultBits));
-              } else {
-                if (toTermSet instanceof BitDocSet) {
-                  resultBits = ((BitDocSet)toTermSet).bits.clone();
+              try {
+                // use filter cache
+                DocSet toTermSet = toSearcher.getDocSet(toDeState);
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/search/JoinQParserPlugin.java:592]: Shouldn't only this line be wrapped in try-catch for IllegalStateException?

Comment (thomaswoeckinger) [reply solr/core/src/java/org/apache/solr/search/JoinQParserPlugin.java:592]: toTermSet is used in whole block, so it would introduce null checks

**Comment by thomaswoeckinger** [comment]
> I'm running this test and it failed the first time (meaning the expected exception wasn't thrown); the second is hanging on me still.

Changed the document size, i run the test about 500 times, on my machine the expected exception was always thrown

**Comment by dsmiley** [comment]
Please don't force-push PRs as it inhibits a reviewers ability to see what was changed.

This time I ran via the CLI instead of IntelliJ and it passed.  Maybe my IntelliJ dependency versions were stale (Caffeine was updated recently) so I re-generated and still same result with IntelliJ.  Weird; is it just me?

I constrained the try-catch around just the call to `toTermSet = toSearcher.getDocSet(toDeState);` and it passed; I don't get your point Thomas about resultList.

The test adds tons of over 3k docs to Solr one-by-one instead of batching.  Not a deal breaker but sad when I see it.  Try this:
```
    SolrClient client = new EmbeddedSolrServer(h.getCore());
    client.add(persons);
```

But hold up...

This logic being added, I noticed, is able to see the filter cache config without SolrCache having isAsync() and therefore I think we can make this thing reproduce every time without actually adding any data for a test.  See my related comment in JIRA.  Wouldn't that be better?  Better to eagerly fail than to have a failure sneak up on you in production.

**Comment by thomaswoeckinger** [comment]
> Please don't force-push PRs as it inhibits a reviewers ability to see what was changed.
> 

Okay, so before merge as squash is done?

> This time I ran via the CLI instead of IntelliJ and it passed. Maybe my IntelliJ dependency versions were stale (Caffeine was updated recently) so I re-generated and still same result with IntelliJ. Weird; is it just me?
> 

I use eclipse, and was not able to fail the test.

> I constrained the try-catch around just the call to `toTermSet = toSearcher.getDocSet(toDeState);` and it passed; I don't get your point Thomas about resultList.

It is not about the resultList; toTermSet is used in the next few lines and therefore it must be defined before the try-catch block. Keeping it in the try block would even allow it to be final (maybe better readability), but anyway this is a detail discussion, if you prefer it I can change it.  
> 
> The test adds tons of over 3k docs to Solr one-by-one instead of batching. Not a deal breaker but sad when I see it. Try this:
> 
> ```
>     SolrClient client = new EmbeddedSolrServer(h.getCore());
>     client.add(persons);
> ```
I will change it.
> 
> But hold up...
> 
> This logic being added, I noticed, is able to see the filter cache config without SolrCache having isAsync() and therefore I think we can make this thing reproduce every time without actually adding any data for a test. See my related comment in JIRA. Wouldn't that be better? Better to eagerly fail than to have a failure sneak up on you in production.

Adding isAsync() to the interface is good idea indeed, but will change the runtime behavior because such queries will always fail, which is currently not the case.
Anyway we have to made a decision.

**Pull request closed without merging**